middleware/policy.es

```diff
@@ -0,0 +1,105 @@
+
+, frames // `frame-src 'self' https://${domain};`
+    = Array.from (defaults)
+
+, connects // `connect-src 'self' https://${domain};`
+    = Array.from (defaults)
+
+, images // `image-src 'self' data: https://cdn.example.com`
+    = Array.from (defaults)
+
+, fonts // `font-src 'self' https://cdn.example.com`
+    = Array.from (defaults)
+
+, objects // `object-src 'self' https://cdn.example.com`
+    = Array.from (defaults)
+
+, medias // `media-src 'self' https://cdn.example.com`
+    = Array.from (defaults)
+
+, styles // `style-src 'self' 'unsafe-inline' https://cdn.example.com`
+    = Array.from (defaults)
+
+, scripts // Script Nonce for inline <script>
+  // https://csp.withgoogle.com/docs/strict-csp.html
+  // scripts-src 'self'
+  // nonce-${nonce}
+  // ** NONCE MUST BE UNIQUE **
+  //   https://w3c.github.io/webappsec-csp/#framework-directive-source-list
+  // **NEVER EXPOSE!!! Causes XSS attacks** script-src 'unsafe-inline'
+  // **THAT BEING SAID...For Safari 😢
+  // 'unsafe-inline' // THIS MAY NOT BE TRUE IN 2018
+    = Array.from (defaults)
+
+, policies
+    = [
+
+    // Reporting directives
+      `report-uri ${ reports.join ` ` }` // backwards compatibility
+//  , `report-to ${ reports.join ` ` }`  // greenfield
+
+    // Fetch directives
+    , `default-src ${ defaults.join ` ` }`
+    , `frame-src ${ frames.join ` ` }`
+    , `connect-src ${ connects.join ` ` }`
+    , `img-src ${ images.join ` ` }`
+    , `font-src ${ fonts.join ` ` }`
+    , `object-src ${ objects.join ` ` }`
+    , `media-src ${ medias.join ` ` }`
+    , `style-src ${ styles.join ` ` }`
+    , `script-src ${ scripts.join ` ` }`
+
+//  , `worker-src ${ workers.join ` ` }`
+
+    // Document directives
+//  , `form-action ${ forms.join ` ` }`
+//  , `frame-ancestors ${ ancestors.join ` ` }`
+//  , `sandbox ${ sandboxes.join ` ` }`
+
+//  // Other directives
+//  // https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity
+//  // https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/require-sri-for
+//  , `require-sri-for ${ integrities.join ` ` }`
+//  , `block-all-mixed-content ${ reports.join ` ` }`
+//  // https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/upgrade-insecure-requests
+//  , `update-insecure-requests`
+
+//  // DEPRECATED!! See https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Referrer-Policy
+//  , `referrer`
+    ]
+
+
+module.exports = options =>
+
+  async (context, next, header) => {
+
+    await next ()
+
+
+    header
+
+      // https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy
+      = 'Content-Security-Policy'
+
+      // Is this a security breach?
+      // Will someone be able to disable CSP with this?
+      // https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy-Report-Only
+      + ( 'report' in context.request.query ? '-Report-Only' : '' )
```

Comment:
@brandondees trying to figure out a way to trigger reporting. I feel having a querystring parameter `?report`
Should only work in development but that's a side effect that now has to be tracked. Apparently can do BOTH
`Content-Security-Policy` and `Content-Security-Policy-Report-Only` and they are complimentary and inconsequential. Perhaps that's our strategy. https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy-Report-Only

Reply:
wow am i behind the times on this. the report only feature looks like something everybody should enable all the time as a general best practice, but there are always exceptions to the rule for this kind of thing. where's this code going to have an effect? just on snuggsi.es? or for all users of the shim? or just on the docs website? i'm trying to gauge how much we're thinking about "show off and prescribe a best practice we figured out" vs "we need this to work right" vs "just exploring what's possible"

both csp and the reporting are things that there might be valid reasons to _not_ use, or to configure differently, naturally. but for typical web developers building apps, we should be able to come up with some sort of sensible default starting point that errs towards more secure rather than less. whitelist by default, blacklist instead only if it fits your model better, disable altogether if you really know what you're doing.

Reply:
@brandondees great question. Basically for now eating own dogfood. (And great fallback reference of "we did this before here's where it's at since it's relevant moving forward". Great statements and thoughts. I'm having the same.

I found this as well. Perhaps `?report` just enables reporting. Should be inconsequential of the explicit `Content-Security-Policy`. And would only report on what is in `Content-Security-Policy-Report-Only`. I feel making them mirror images of each other ALWAYS would at least provide more Occam's razor and less Murphy's Law.

Thoughts...

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy#Multiple_content_security_policies

Reply:
P.S. @brandondees better get up on it. YUGE part of Rails 5.2.